### PR TITLE
Conserve cause of `ImplDerivedObligation` in E0599

### PR DESCRIPTION
**Pull request by estebank**
CC #86377.

### CODE REVIEW

File: src/test/ui/higher-rank-trait-bounds/issue-30786.stderr [hunk 1]
```diff
@@ -10,14 +10,21 @@ LL | pub struct Map<S, F> {
...
+note: trait bound `&'a mut &mut Map<Repeat, [closure@$DIR/issue-30786.rs:117:27: 117:34]>: Stream` was not satisfied
+  --> $DIR/issue-30786.rs:96:50
+   |
+LL | impl<T> StreamExt for T where for<'a> &'a mut T: Stream {}
+   |         ---------     -                          ^^^^^^ unsatisfied trait bound `for<'a> &'a mut &mut Map<Repeat, [closure@issue-30786.rs:117:27]>: Stream` introduced here
+note: trait bound `&'a mut Map<Repeat, [closure@$DIR/issue-30786.rs:117:27: 117:34]>: Stream` was not satisfied
+  --> $DIR/issue-30786.rs:96:50
+   |
+LL | impl<T> StreamExt for T where for<'a> &'a mut T: Stream {}
+   |         ---------     -                          ^^^^^^ unsatisfied trait bound `for<'a> &'a mut Map<Repeat, [closure@issue-30786.rs:117:27]>: Stream` introduced here
```
Comment (estebank) [inline src/test/ui/higher-rank-trait-bounds/issue-30786.stderr:27]: I need to revert this change

File: src/test/ui/suggestions/derive-trait-for-method-call.stderr [hunk 2]
```diff
@@ -16,10 +16,27 @@ LL | struct Foo<X, Y> (X, Y);
...
+   |                 ^^^^^^^                         ---------
+   |                 |
+   |                 unsatisfied trait bound introduced here
+note: trait bound `CloneEnum: Default` was not satisfied
+  --> $DIR/derive-trait-for-method-call.rs:20:40
+   |
+LL | impl<X: Clone + Default + , Y: Clone + Default> Foo<X, Y> {
+   |                                        ^^^^^^^  ---------
+   |                                        |
+   |                                        unsatisfied trait bound introduced here
```
Comment (compiler-errors) [inline src/test/ui/suggestions/derive-trait-for-method-call.stderr:39]: r=me with or without logic to flatten these into one note :heart:

**Review by compiler-errors — APPROVED** [review]

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 3]
```diff
@@ -1556,7 +1556,23 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
 
                     // Convert the bounds into obligations.
                     let impl_obligations = traits::predicates_for_generics(
-                        |_, _| cause.clone(),
+                        |_idx, span| {
+                            let misc = traits::ObligationCause::misc(span, self.body_id);
+                            let parent_trait_pred = ty::Binder::dummy(ty::TraitPredicate {
+                                trait_ref: ty::TraitRef::from_method(self.tcx, impl_def_id, substs),
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/probe.rs:1562]: I should've caught this when reviewing this, but this is not correct. We cannot construct a trait ref out of an impl def id and impl substs, and it's not meaningful to even use `tcx.impl_trait_ref(impl_def_id).subst(self.tcx, substs)` here because this is an *inherent* impl by construction.

I fixed this in #110133
